linkcheck/checker/fileurl.py

```diff
@@ -314,3 +305,73 @@ def add_url(self, url, line=0, column=0, page=0, name="", base=None):
             url = webroot + url[1:]
             log.debug(LOG_CHECK, "Applied local webroot `%s' to `%s'.", webroot, url)
         super().add_url(url, line=line, column=column, page=page, name=name, base=base)
+
+
+class AnchorCheckFileUrl(FileUrl):
+    def reset(self):
+        super().reset()
+        # the local file URI
+        self.url_without_anchor = None
+
+    def build_url(self):
+        """
+        Calls UrlBase.build_url() and adds a trailing slash to directories.
+        """
+        self.build_base_url()
+        if self.parent_url is not None:
+            # URL joining with the parent URL only works if the query
+            # of the base URL are removed first.
+            # Otherwise the join function thinks the query is part of
+            # the file name.
+            urlparts = list(urllib.parse.urlsplit(self.base_url))
+            # ignore query part for filesystem urls
+            urlparts[3] = ''
+            self.base_url = urlutil.urlunsplit(urlparts)
+        super(FileUrl, self).build_url()
```

Comment:
I see what you're doing, and it seems like a reasonable compromise for the short term. But things like this should be a sign that something is still off. So far, I can't envision any particular problem it might cause.